lib/quicly.c

```diff
@@ -408,27 +414,37 @@ static const quicly_transport_parameters_t default_transport_params = {.max_udp_
                                                                        .max_ack_delay = QUICLY_DEFAULT_MAX_ACK_DELAY,
                                                                        .min_ack_delay_usec = UINT64_MAX};
 
-static __thread int64_t now;
-static __thread size_t in_quicly_send_cnt;
+struct st_quicly_lock_now_state_t {
+    uint8_t unlock_on_exit; /* boolean */
+};
 
-static void update_now(quicly_context_t *ctx)
+static void lock_now(quicly_conn_t *conn, struct st_quicly_lock_now_state_t *state)
 {
-    if (in_quicly_send_cnt == 0) {
-        int64_t newval = ctx->now->cb(ctx->now);
-        if (now < newval)
-            now = newval;
+    if (conn->stash.now != 0) {
+        assert(state != NULL && "caller is not reentrant");
+        state->unlock_on_exit = 0;
+    } else {
+        if (state != NULL)
+            state->unlock_on_exit = 1;
+
+        /* update _now, but never let it go back */
+        conn->stash.now = conn->super.ctx->now->cb(conn->super.ctx->now);
+        static __thread int64_t prev_now;
+        if (conn->stash.now < prev_now) {
```

Comment:
How can this happen?

Reply:
As an example, when the now callback is referring to system clock, and if the system clock is rewinded to the past.